### PR TITLE
allow env vars in place of flags

### PR DESCRIPTION
**Pull request by GraemeChorusOne**

### CODE REVIEW

**Comment by joe-bowman** [comment]
@YunSuk-Yeo - this replaces #31 with the previous suggestion applied, and brought into line with the latest version.

File: feeder/src/vote.ts [hunk 1]
```diff
@@ -296,19 +296,20 @@ interface VoteArgs {
...
-    chainID: args.chainID,
-    gasPrices: args.gasPrices,
+    URL: args.lcdAddress || process.env['LCD_ADDRESS'],
+    chainID: args.chainID || process.env['CHAIN_ID'],
+    gasPrices: args.gasPrices || process.env['GAS_PRICE'],
   })
-  const rawKey: RawKey = await initKey(args.keyPath, args.password)
-  const valAddrs = args.validator || [rawKey.valAddress]
+  const rawKey: RawKey = await initKey(process.env['KEY_PATH'] || args.keyPath, process.env['PASSPHRASE'] || args.password)
+  const valAddrs = (process.env['VALIDATOR'] && process.env['VALIDATOR'].split(',')) || args.validator || [rawKey.valAddress]
```
Comment (kucharskim) [inline feeder/src/vote.ts:304]: I think `args` should have higher priority than `env`. See my comments in #31. Priority of `args` vs `env` is applicable to all changes across the codebase. So whatever is done in this pull request needs to have order of `args` first, `env` second.

File: feeder/src/vote.ts [hunk 2]
```diff
@@ -296,19 +296,20 @@ interface VoteArgs {
...
   const voterAddr = rawKey.accAddress
   const wallet = new Wallet(client, rawKey)
+  const source = (process.env['SOURCE'] && process.env['SOURCE'].split(',')) || args.source
 
   while (true) {
     const startTime = Date.now()
 
-    await processVote(client, wallet, args.source, valAddrs, voterAddr).catch((err) => {
+    await processVote(client, wallet, source, valAddrs, voterAddr).catch((err) => {
```
Comment (kucharskim) [inline feeder/src/vote.ts:312]: See my comment in #31 about `required: false`. How this tool will behave, if any of currently required arguments will not be provided via `args` or vie `env`?

**Review by kucharskim — CHANGES_REQUESTED** [review]

File: feeder/src/vote.ts [hunk 3]
```diff
@@ -296,19 +296,20 @@ interface VoteArgs {
 
 export async function vote(args: VoteArgs): Promise<void> {
   const client = new LCDClient({
-    URL: args.lcdAddress,
-    chainID: args.chainID,
-    gasPrices: args.gasPrices,
+    URL: args.lcdAddress || process.env['LCD_ADDRESS'],
+    chainID: args.chainID || process.env['CHAIN_ID'],
+    gasPrices: args.gasPrices || process.env['GAS_PRICE'],
```
Comment (kucharskim) [inline feeder/src/vote.ts:301]: I'm not sure will this work as `-g` has `defaultValue` set.

File: feeder/src/vote.ts [hunk 4]
```diff
@@ -295,20 +295,34 @@ interface VoteArgs {
...
-  const client = new LCDClient({
-    URL: args.lcdAddress,
-    chainID: args.chainID,
-    gasPrices: args.gasPrices,
-  })
-  const rawKey: RawKey = await initKey(args.keyPath, args.password)
-  const valAddrs = args.validator || [rawKey.valAddress]
+  let lcdAddress:string = args.lcdAddress || process.env['LCD_ADDRESS'] || ''
+  let chainID:string    = args.chainID    || process.env['CHAIN_ID']    || ''
+  let gasPrices:string  = args.gasPrices  || process.env['GAS_PRICE']   || ''
```
Comment (joe-bowman) [inline feeder/src/vote.ts:300]: This won't work for gasPrices - as the default is set in the argParser.

We'd have to move the default from the argParser to the fallback argument here.

File: feeder/src/vote.ts (hunk 2 above)
Comment (GraemeChorusOne) [reply feeder/src/vote.ts:325]: Updated with check, if neither are set then they will be set to an empty string.  Then check for the empty string.

File: feeder/src/vote.ts (hunk 1 above)
Comment (GraemeChorusOne) [reply feeder/src/vote.ts]: Have made the swap

File: feeder/src/vote.ts [hunk 5]
```diff
@@ -295,20 +295,34 @@ interface VoteArgs {
...
+  {
+    console.error('Missing either --keypath or --password')
+    return
+  }
+  const rawKey: RawKey = await initKey(keyPath, passphrase)
+
+  const valAddrs = args.validator || (process.env['VALIDATOR'] && process.env['VALIDATOR'].split(',')) || [rawKey.valAddress]
   const voterAddr = rawKey.accAddress
   const wallet = new Wallet(client, rawKey)
+  const source = (process.env['SOURCE'] && process.env['SOURCE'].split(',')) || args.source
```
Comment (joe-bowman) [inline feeder/src/vote.ts:320]: This needs to be reverted so `--source` takes precedence.

File: feeder/src/vote.ts [hunk 6]
```diff
@@ -295,20 +295,34 @@ interface VoteArgs {
...
+  let chainID:string    = args.chainID    || process.env['CHAIN_ID']    || ''
+  let gasPrices:string  = args.gasPrices  || process.env['GAS_PRICE']   || ''
+  if (lcdAddress == '' || chainID == '' || gasPrices == '')
+  {
+    console.error('Missing --lcd, --chain or --gas-prices')
+    return
+  }
+  const client = new LCDClient({ URL: lcdAddress, chainID: chainID, gasPrices: gasPrices, })
+
+  let keyPath:string    = args.keyPath || process.env['KEY_PATH'] || ''
```
Comment (joe-bowman) [inline feeder/src/vote.ts:308]: Another default value, same action here.

**Review by joe-bowman — CHANGES_REQUESTED** [review]

File: feeder/src/vote.ts [hunk 7]
```diff
@@ -295,20 +295,34 @@ interface VoteArgs {
...
-  const client = new LCDClient({
-    URL: args.lcdAddress,
-    chainID: args.chainID,
-    gasPrices: args.gasPrices,
-  })
-  const rawKey: RawKey = await initKey(args.keyPath, args.password)
-  const valAddrs = args.validator || [rawKey.valAddress]
+  let lcdAddress:string = args.lcdAddress || process.env['LCD_ADDRESS'] || ''
+  let chainID:string    = args.chainID    || process.env['CHAIN_ID']    || ''
+  let gasPrices:string  = args.gasPrices  || process.env['GAS_PRICE']   || '169.77ukrw'
```
Comment (YunSuk-Yeo) [inline feeder/src/vote.ts:300]: can we remove gasPrices?

now oracle votes does not use gas prices

File: feeder/src/index.ts [hunk 8]
```diff
@@ -57,14 +57,14 @@ function registerCommands(parser: ArgumentParser): void {
     action: `store`,
     help: `key store path to save encrypted key`,
     dest: `keyPath`,
-    defaultValue: `voter.json`,
+    required: false,
   })
 
   voteCommand.addArgument([`-g`, `--gas-prices`], {
```
Comment (YunSuk-Yeo) [inline feeder/src/index.ts:63]: we can remove gas prices, not used anymore

**Review by YunSuk-Yeo — CHANGES_REQUESTED** [review]

**Comment by YunSuk-Yeo** [comment]
replaced by #50

**Pull request closed without merging**